refactor(subscriptions): use a single payment update form

chenba opened the pull request:
Because:
 - payment method for subscriptions should be updated in one place

This commit:
 - move the payment method info and update form into their own section
 - move CancelSubscriptionPanel into its own file
 - replace 'react-test-renderer' tests with 'Testing-Library/react' ones
 - make minor style and layout changes
 - remove product and plan ids from payment update Amplitude events as
   they are no longer associated with a subscription


## Issue that this pull request solves

Closes: FXA-2316

## Checklist

_Put an `x` in the boxes that apply_

- [x] My commit is GPG signed.
- [x] If applicable, I have modified or added tests which pass locally.
- [ ] I have added necessary documentation (if appropriate).

## Screenshots 
![image](https://user-images.githubusercontent.com/198055/89428211-c055a980-d701-11ea-9984-692711dda3f4.png)

chenba commented on packages/fxa-payments-server/src/lib/test-utils.tsx:
```diff
@@ -509,6 +509,7 @@ export const MOCK_CUSTOMER = {
       cancel_at_period_end: false,
       current_period_start: 1565816388.815,
       current_period_end: 1568408388.815,
+      end_at: null,
```
This fixes a type issue.

chenba commented on packages/fxa-payments-server/src/routes/Subscriptions/Cancel/CancelSubscriptionPanel.test.tsx:
```diff
@@ -0,0 +1,212 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
This file replaces the rendering tests using react-test-renderer that were in `src/routes/Subscriptions/PaymentUpdateForm.test.tsx`, as well as adding a couple event handling tests that were previously missing.

chenba commented on packages/fxa-payments-server/src/routes/Subscriptions/Cancel/CancelSubscriptionPanel.tsx:
```diff
@@ -0,0 +1,196 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
I moved this component from SubscriptionItem into its own file.  It's a little more consistent since Reactivate has its own dir and all.  It's also a little easier to find this file or the tests by having 'cancel' in the filenames.

chenba commented on packages/fxa-payments-server/src/routes/Subscriptions/PaymentUpdateForm.test.tsx:
```diff
@@ -1,318 +1,56 @@
+  updatePayment: jest.fn().mockResolvedValue(null),
+  updatePaymentStatus: defaultState.updatePayment,
+};
+
```
As mentioned above (if you are viewing this on GitHub), the tests have been replaced and moved: https://github.com/mozilla/fxa/pull/6144/files#r465808631

I added a couple simple rendering tests below.  Keep in mind that:
 - the payment form has its own tests in `src/components/PaymentForm/index.test.tsx`
 - we are planning to remove all the /v1 code, so expending effort to add a lot of tests or even refactor /v1 code to make those tests possible is wasteful

chenba commented on packages/fxa-payments-server/src/routes/Subscriptions/index.tsx:
```diff
@@ -293,6 +300,32 @@ export const Subscriptions = ({
             </div>
           </div>
 
+          {customer.result && showPaymentUpdateForm && (
```
TS's typechecker is not good enough to know that `showPaymentUpdateForm` already contains a check on `customer.result`.  🤷‍♂️

lmorchard replied:
Hmm, probably a bad smell that could be improved in issue #5683, but :shrug: for now

lmorchard commented on packages/fxa-payments-server/src/routes/Subscriptions/SubscriptionItem.tsx:
```diff
@@ -139,154 +81,4 @@ export const SubscriptionItem = ({
-  cancelSubscriptionStatus: SelectorReturns['cancelSubscriptionStatus'];
-};
-
-const CancelSubscriptionPanel = ({
```
:+1: Nice to see this component really slim down with the extraction of CancelSubscriptionPanel

lmorchard commented on packages/fxa-payments-server/src/store/actions/api.ts:
```diff
@@ -83,14 +83,11 @@ export default {
         };
       },
     } as const),
-  updatePayment: (paymentToken: string, plan: Plan) =>
```
Another spot that probably needs replication for the SCA side of payment update, since I added metrics stuff to it.

chenba commented on packages/fxa-payments-server/src/lib/amplitude.ts:
```diff
@@ -160,31 +160,27 @@ export function createSubscription_REJECTED(eventProperties: EventProperties) {
-    eventTypeNames.view,
-    eventProperties
-  );
+export function updatePaymentMounted(eventProperties: EventProperties = {}) {
```
I waffled a bit on this since Amplitude event properties can be any key-value pairs.  But since we are not really using the param, it might be confusing to leave it there.  I'll remove it.

chenba commented:
Rebased on main, and pushed the following changes to address review feedback:
 - removed the Amplitude event properties param that was only used to pass the product and plan ids on payment updates, including on the new SCA functions
 - added a test to `CancelSubscriptionPanel.test.tsx` to assert that we log the two Amplitude metrics events ("viewed" and "engaged")
